nexus/db-queries/src/db/datastore/vpc.rs

```diff
@@ -1616,12 +2276,125 @@ impl DataStore {
             })
     }
 
+    /// Fetch all custom routers in a VPC.
+    pub async fn vpc_get_custom_routers(
+        &self,
+        opctx: &OpContext,
+        vpc_id: Uuid,
+    ) -> ListResultVec<VpcRouter> {
+        use db::schema::vpc_router::dsl as router_dsl;
+
+        router_dsl::vpc_router
+            .filter(router_dsl::time_deleted.is_null())
+            .filter(router_dsl::vpc_id.eq(vpc_id))
+            .select(VpcRouter::as_select())
+            .load_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Vpc,
+                        LookupType::ById(vpc_id),
+                    ),
+                )
+            })
+    }
+
+    // XXX: maybe wants to live in external IP?
+    // XXX: this assumes that only one IGW can be associated
+    //      with each IP per VPC. I would be... surprised if
+    //      we allow a pool to be referenced by several IGWs
+    //      in the same VPC? But maybe that's worth discussion.
+    //      There's nothing stopping us from changing OPTE to
+    //      insert several nat rules if there exists >1 IGW value
+    //      (or just implementing an OR predicate).
+    /// Returns a (Ip, NicId) -> InetGwId map which identifies, for a given sled:
+    /// * a) which external IPs belong to any of its services/instances/probe NICs.
+    /// * b) whether each IP is linked to an Internet Gateway via its parent pool.
+    pub async fn vpc_resolve_sled_external_ips_to_gateways(
+        &self,
+        opctx: &OpContext,
+        sled_id: Uuid,
+    ) -> Result<HashMap<Uuid, HashMap<IpAddr, Uuid>>, Error> {
+        // TODO: give GW-bound addresses preferential treatment.
```

Comment:
This should be implemented now. It's operating via the same mechanism that we're using to correctly handle multiple IGWs linking against a single pool, rather than assigning priority to single IP mappings.